### PR TITLE
HADOOP-16251. ABFS: add FSMainOperationsBaseTest

### PR DESCRIPTION
**Pull request by DadanielZ**
- Addded tests which are extended from FSMainOperationsBaseTest

### CODE REVIEW

**Comment by DadanielZ** [comment]
Hey @steveloughran what should I do to gain the access to label the PR?

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1032 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 34 | trunk passed |
| +1 | shadedclient | 746 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 41 | trunk passed |
| +1 | javadoc | 23 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 26 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| -0 | checkstyle | 13 | hadoop-tools/hadoop-azure: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 25 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 742 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 46 | the patch passed |
| +1 | javadoc | 20 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 81 | hadoop-azure in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3042 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-787/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/787 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 26afe97d4ec2 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / dead9b4 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/1/testReport/ |
| Max. process+thread count | 442 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
I just hit the label button. Now, because of your commit rights on the abfs branch, you do get access to all this stuff. I'll send you a link

**Comment by steveloughran** [comment]
code LGTM, as usual for the diligence: which endpoint did you test with?

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  public void testListStatusThrowsExceptionForUnreadableDir() {
+    // Permission Checks:
+    // https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HdfsPermissionsGuide.html
+  }
+
+  @Override
+  @Ignore("There shouldn't be permission check for getFileInfo")
+  public void testGlobStatusThrowsExceptionForUnreadableDir() {
+    // Permission Checks:
+    // https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HdfsPermissionsGuide.html
```
Comment (ajfabbri) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java]: Good patch overall but I commented on the [JIRA](https://issues.apache.org/jira/browse/HADOOP-16251) about these permissions details.

Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java]: Thanks for clarifying this, then it looks like a bug in service : (.

**Comment by DadanielZ** [comment]
@steveloughran Thanks for sharing the link.
I just find there are resource sharing in the test suite which prevent the parallel running, so I updated the setup so each single test can be ran in parallel.

Tests passed against endpoint in west-us:
xns account:
Tests run: 40, Failures: 0, Errors: 0, Skipped: 0
Tests run: 383, Failures: 0, Errors: 0, Skipped: 25
Tests run: 190, Failures: 0, Errors: 0, Skipped: 23

non-xns account:
Tests run: 40, Failures: 0, Errors: 0, Skipped: 0
Tests run: 383, Failures: 0, Errors: 0, Skipped: 207
Tests run: 190, Failures: 0, Errors: 0, Skipped: 15

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 27 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1033 | trunk passed |
| +1 | compile | 24 | trunk passed |
| +1 | checkstyle | 20 | trunk passed |
| +1 | mvnsite | 30 | trunk passed |
| +1 | shadedclient | 681 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 18 | trunk passed |
| 0 | spotbugs | 50 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 49 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 27 | the patch passed |
| +1 | compile | 26 | the patch passed |
| +1 | javac | 26 | the patch passed |
| -0 | checkstyle | 16 | hadoop-tools/hadoop-azure: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 28 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 774 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 19 | the patch passed |
| +1 | findbugs | 61 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 79 | hadoop-azure in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3029 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-787/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/787 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 2df128cb2fd4 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d6b7609 |
| Default Java | 1.8.0_191 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/2/testReport/ |
| Max. process+thread count | 465 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-787/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java (hunk 1 above)
Comment (ajfabbri) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java]: I was wrong. I misread the namenode code. Though it handles an AccessControlException apparently it doesn't check permissions, just HA status.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    return fSys;
+  }
+
+  @Override
+  @Ignore("Permission check for getFileInfo doesn't match the HdfsPermissionsGuide")
+  public void testListStatusThrowsExceptionForUnreadableDir() {
+    // Permission Checks:
+    // https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-hdfs/HdfsPermissionsGuide.html
+  }
+
```
Comment (ajfabbri) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java]: The Ignore annotation says getFileInfo.. I would have thought you wanted the getListing row in the permission chart?

```getListing	NO	N/A	N/A	READ, EXECUTE```

which is READ | EXECUTE on the final part of the path (dir). T thought that was what this test case was checking? I'm +1 on the patch but you should update the Ignore comment if it is indeed getListing that should be checked here. If this behavioir is unique to the ABFS service that's ok we should just document it as such.

**Comment by ajfabbri** [comment]
Committed to trunk. Thank you for the contribution @DadanielZ.  If you'd like to change the comments in to address my open question above feel free to publish another patch.

**Comment by ajfabbri** [comment]
Don't see a way to close this PR. Doesn't look like I have github write access yet. "Only those with write access to this repository can merge pull requests."

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java (hunk 2 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMainOperation.java]: @ajfabbri yes it is a little bit confused here. As the actual call is ListStatus(), which returns FileStatus array. I'm wondering if the permission guide is for the FS level API?

**Pull request closed without merging**